### PR TITLE
Feature/trapz sorted

### PR DESCRIPTION
**Pull request by seajane**
Adding a wrapper function, `trapz_sorted` to automatically sort data for more consistent results when estimating the area under a curve.  Added a `na.rm` parameter to remove NA values in the data.

### CODE REVIEW

File: R/trapz_sorted.R [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+#'
+#' @export
+#'
+trapz_sorted <- function(x, y, na.rm = TRUE){
+  # Input Checking
+  .check_numeric_input(x)
+  .check_numeric_input(y)
+  m <- length(x)
+  if (length(y) != m) stop("Arguments 'x', 'y' must be vectors of the same length.")
+  if (na.rm == TRUE){
```
Comment (wfulp) [inline R/trapz_sorted.R:47]:
```suggestion
  if (na.rm == TRUE) {
```

File: R/trapz_sorted.R [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    # Sort by increasing values, NA are removed from x
+    x <- x[order(x, decreasing = FALSE, method = "auto", na.last = NA)]
+    y <- y[order(x, decreasing = FALSE, method = "auto", na.last = NA)]
+    # NA removed from y
+    x <- x[!is.na(y)]
+    y <- y[!is.na(y)]
+  } else{
+    # Warn that data points will be removed if NA
+    if(anyNA(x)) warning("NA values are present in the 'x' values. Returns NA.")
+    if(anyNA(y)) warning("NA values are present in the 'y' values. Returns NA.")
```
Comment (wfulp) [inline R/trapz_sorted.R:57]:
```suggestion
    if (anyNA(x)) warning("NA values are present in the 'x' values. Returns NA.")
    if (anyNA(y)) warning("NA values are present in the 'y' values. Returns NA.")
```
spacing

File: R/trapz_sorted.R [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+#' set.seed(93)
+#'
+#' # Calculate the area under the sine curve from 0 to pi:
+#' n <- 101
+#' x <- seq(0, pi, len = n)
+#' y <- sin(x)
+#' trapz_sorted(x, y)
+#'
+#' # Calculate area under the curve for unsorted data:
+#' x <- runif(n, 1,42)
```
Comment (wfulp) [inline R/trapz_sorted.R:28]:
```suggestion
#' x <- sample(1:n, n, replace = F)
```

File: R/trapz_sorted.R [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+#' y <- runif(n, 0, 33)
+#' trapz_sorted(x, y)
+#'
+#' # Calculate the area without NA removal
+#' y[3] <- NA
```
Comment (wfulp) [inline R/trapz_sorted.R:33]:
```suggestion
#' x[10] <- NA
#' y[3] <- NA
```

File: R/trapz_sorted.R [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+#' # Will fail to produce a result
+#' trapz_sorted(x, y, na.rm=FALSE)
```
Comment (wfulp) [inline R/trapz_sorted.R:35]:
```suggestion
#' trapz_sorted(x, y, na.rm=FALSE)
#' trapz_sorted(x, y, na.rm=TRUE)
```

File: R/trapz_sorted.R [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+#'
+#'
+#'
+#' @export
+#'
+trapz_sorted <- function(x, y, na.rm = TRUE){
+  # Input Checking
+  .check_numeric_input(x)
```
Comment (wfulp) [inline R/trapz_sorted.R:43]: I think we need a check for duplicate values of x (probably throwing an error). Currently the results are unexpected (uses the y value of the last x in the dataset)

**Review by wfulp — COMMENTED** [review]
Great Addition. I added some minor changes. One was changing x in examples to be closer to our normal dilution curve data.

File: R/trapz_sorted.R (hunk 6 above)
Comment (seajane) [reply R/trapz_sorted.R:45]: added an error if there are duplicated x values https://github.com/FredHutch/VISCfunctions/pull/66/commits/04255c56537fd977aa1da48a0312c4f6c3ca66bd

**Review by wfulp — APPROVED** [review]
Wahoo ready to merge in.....looks great

**Pull request merged**